pq-crypto/s2n_kyber_512_evp.c

```diff
@@ -0,0 +1,87 @@
+/*
+* Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+*
+* Licensed under the Apache License, Version 2.0 (the "License").
+* You may not use this file except in compliance with the License.
+* A copy of the License is located at
+*
+*  http://aws.amazon.com/apache2.0
+*
+* or in the "license" file accompanying this file. This file is distributed
+* on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
+* express or implied. See the License for the specific language governing
+* permissions and limitations under the License.
+*/
+
+#include "s2n_kyber_512_evp.h"
+
+#include <openssl/evp.h>
+#include <stddef.h>
+
+#include "error/s2n_errno.h"
+#include "tls/s2n_kem.h"
+#include "utils/s2n_safety_macros.h"
+
+#if defined(AWS_LC_SUPPORTS_KYBER_512)
+int s2n_kyber_512_evp_generate_keypair(unsigned char *public_key, unsigned char *private_key) {
+    EVP_PKEY_CTX *kyber_pkey_ctx = EVP_PKEY_CTX_new_id(EVP_PKEY_KYBER512, NULL);
+    POSIX_GUARD_PTR(kyber_pkey_ctx);
+    POSIX_ENSURE(EVP_PKEY_keygen_init(kyber_pkey_ctx), S2N_FAILURE);
+
+    EVP_PKEY *kyber_pkey = NULL;
+    POSIX_ENSURE(EVP_PKEY_keygen(kyber_pkey_ctx, &kyber_pkey), S2N_FAILURE);
+    POSIX_GUARD_PTR(kyber_pkey->pkey.ptr);
+
+    size_t public_key_size = S2N_KYBER_512_R3_PUBLIC_KEY_BYTES;
+    size_t private_key_size = S2N_KYBER_512_R3_SECRET_KEY_BYTES;
+    POSIX_ENSURE(EVP_PKEY_get_raw_public_key(kyber_pkey, public_key, &public_key_size), S2N_FAILURE);
+    POSIX_ENSURE(EVP_PKEY_get_raw_private_key(kyber_pkey, private_key, &private_key_size), S2N_FAILURE);
+
+    return S2N_SUCCESS;
+}
+
+int s2n_kyber_512_evp_encapsulate(unsigned char *ciphertext, unsigned char *shared_secret,
+                                  const unsigned char *public_key) {
+    size_t public_key_size = S2N_KYBER_512_R3_PUBLIC_KEY_BYTES;
+    EVP_PKEY *kyber_pkey = EVP_PKEY_new_raw_public_key(EVP_PKEY_KYBER512, NULL, public_key, public_key_size);
+    POSIX_GUARD_PTR(kyber_pkey);
+
+    EVP_PKEY_CTX *kyber_pkey_ctx = EVP_PKEY_CTX_new(kyber_pkey, NULL);
+    POSIX_GUARD_PTR(kyber_pkey_ctx);
+
+    size_t cipher_text_size = S2N_KYBER_512_R3_CIPHERTEXT_BYTES;
```

Comment:
Just an FYI that these values can be retrieved dynamically from `EVP_PKEY_encapsulate()` using `ciphertext = NULL`. Current method is faster of course. Dunno whether these values would change post standardisation.

Reply:
We're on the same page. My rationale here was that these values are part of a specification so I chose to use a compile time constant rather than to query it at runtime for performance reasons. We also depend on the compile time constants elsewhere (e.g. in the s2n_kem struct) so I figured consistency is necessary.

What's our general approach around developing around the moving goal posts of standardization?

Reply:
IMO, I see It as inherently experimental until a finalised standard is in place. Not sure how this PQ layer has been built to support a potential instable standard or what the plan is. @alexw91 might be able to expand.